src/facade/reply_builder.h

```diff
@@ -167,54 +167,119 @@ class SinkReplyBuilder {
     }
 
-   private:
-    bool prev_scoped;
-    SinkReplyBuilder2* rb;
+    ~ReplyAggregator();
   };
 
- public:
-  void Write(std::string_view str);
+  void Flush();  // Send all accumulated data and reset to clear state
+
+  std::error_code GetError() const {
+    return ec_;
+  }
+
+  size_t UsedMemory() const {
+    return buffer_.Capacity();
+  }
+
+  bool IsSendActive() {
+    return send_active_;
+  }
+
+  void SetBatchMode(bool b) {
+    batched_ = b;
+  }
+
+  void CloseConnection();
+
+  static const ReplyStats& GetThreadLocalStats() {
+    return tl_facade_stats->reply_stats;
+  }
+
+  static void ResetThreadLocalStats() {
+    tl_facade_stats->reply_stats = {};
+  }
+
+ public:  // High level interface
+  virtual void SendLong(long val) = 0;
+  virtual void SendSimpleString(std::string_view str) = 0;
+
+  virtual void SendStored() = 0;
+  virtual void SendSetSkipped() = 0;
+  void SendOk() {
+    SendSimpleString("OK");
+  }
+
+  virtual void SendError(std::string_view str, std::string_view type = {}) = 0;  // MC and Redis
+  void SendError(OpStatus status);
+  void SendError(ErrorReply error);
+  virtual void SendProtocolError(std::string_view str) = 0;
 
  protected:
-  void Flush();        // Send all accumulated data and reset to clear state
+  // Write strings beyond the inline size as refs. Once the scope ends, they're flushed
+  // or converted into pieces if batching is enabled and the buffer is not filled up enough.
+  void WriteI(std::string_view str) {
+    str.size() > kMaxInlineSize ? WriteRef(str) : WritePiece(str);
+  }
+
+  // Constexpr arrays are assumed to be protocol control sequences, stash them as pieces
+  template <size_t S> void WriteI(const char (&arr)[S]) {
+    WritePiece(std::string_view{arr, S - 1});  // we assume null termination
+  }
+
+  template <typename... Args> void Write(Args&&... strs) {
+    (WriteI(strs), ...);
+  }
+
   void FinishScope();  // Called when scope ends
 
   char* ReservePiece(size_t size);        // Reserve size bytes from buffer
   void CommitPiece(size_t size);          // Mark size bytes from buffer as used
   void WritePiece(std::string_view str);  // Reserve + memcpy + Commit
 
   void WriteRef(std::string_view str);  // Add iovec bypassing buffer
-
-  bool IsInBuf(const void* ptr) const;  // checks if ptr is part of buffer_
   void NextVec(std::string_view str);
 
  private:
   io::Sink* sink_;
   std::error_code ec_;
 
-  bool scoped_;
+  bool send_active_ = false;
+  bool scoped_ = false, batched_ = false;
 
-  size_t total_size_ = 0;  // sum of vec_ lengths
-  base::IoBuf buffer_;
-  std::vector<iovec> vecs_;
+  size_t total_size_ = 0;                          // sum of vec_ lengths
+  base::IoBuf buffer_;                             // backing buffer for pieces
+  absl::InlinedVector<iovec, 16> vecs_;            // vectors for writev
+  absl::InlinedVector<unsigned, 16> ref_indices_;  // indices of ref vectors
```

Comment:
I did not understand what purpose it serves.

Reply:
Line 248 has a comment, you can help me phrase it more clearly. 

We do not copy long strings, but instead refernce them just as vectors to vec_. Once their lifetime ends (i.e. the scope guards drops), we can either send them or keep them to continue batching. If we want to keep them, we have to copy the "ref" data to the buffer and swap references

Reply:
hmm, why would you want to keep them? imagine sending a 1MB string, for example.

Reply:
Yes, but we reference strings starting from 32 bytes. If we stored references to a few 32 byte strings, it would make sense to copy them to the buffer to minimize send calls. We can use another inline limit, we can use other constants, but I assume there will be a batch size that is more efficient to copy than send in two calls